text/0031-transactions-with-contingent-inputs.md

```diff
@@ -0,0 +1,408 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+To support mixed transactions, the `Tx` format is extended as follows.
+
+The `SignatureRctBulletProofs` object has two new fields, indicating token ids of outputs and pseudo output commitments.
+
+```
+message SignatureRctBulletproofs {
+    repeated RingMLSAG ring_signatures = 1;
+    repeated CompressedRistretto pseudo_output_commitments = 2;
+    bytes range_proof = 3;
+    repeated bytes range_proofs = 4;
+    repeated fixed32 pseudo_output_commitment_token_ids = 4;
+    repeated fixed32 output_token_ids = 5;
+}
+```
+
+The `pseudo_output_commitment_token_ids` list must be the same length as `pseudo_output_commitments`,
+and `output_token_ids` must be the same length as `outputs` in the `TxPrefix`, or the transaction is invalid.
+(Alternatively, for backwards compatibility we could allow that if these lists are empty then they are filled with zeros.)
+
+We also need to add a `range_proofs` field to accommodate mixed transactions. This is because the current implementation
+of `bulletproofs` which we get from Dalek crate only allows one generator to be used for range proof aggregation. However,
+we believe that upstream could add support for using multiple generators in a single range proof, and then this complexity
+could be removed.
+
+In the `TxPrefix`, the `token_id` field is renamed to `fee_token_id`, and it only represents the token id of
+the fee output now.
+
+```
+// A transaction that a client submits to consensus
+message TxPrefix {
+    // Transaction inputs.
+    repeated TxIn inputs = 1;
+
+    // Transaction outputs.
+    repeated TxOut outputs = 2;
+
+    // Fee paid to the foundation for this transaction
+    uint64 fee = 3;
+
+    // The block index at which this transaction is no longer valid.
+    uint64 tombstone_block = 4;
+
+    // Token id for this transaction
+    fixed32 fee_token_id = 5;
+}
+```
+
+When checking range-proofs of outputs and pseudo-outputs, we must use the generator indicated by `pseudo_output_commitment_token_ids`
+and `output_token_ids`, rather than the `token_id` in the `TxPrefix`.
+
+To support contingent inputs, we introduce the following objects:
+
+An `InputRules` object is a protobuf with the following schema:
+
+```
+message InputRules {
+    // Outputs that must appear in the transaction for a transaction using this input to be valid
+    repeated external.TxOut required_outputs = 1;
+    // A maximum tombstone block value for the transaction. If zero, it means no maximum is enforced.
+    fixed64 max_tombstone_block = 2;
+}
+```
+
+It may be hashed in a standard way using the `mc-crypto-digestible` hashing scheme (like everything else).
+
+This `required_outputs` are there to allow the main use-case of agreeing to a swap.
+The `max_tombstone_block` is there both, to allow that some of the required outputs can be fog outputs (since fog imposes tombstone block limits),
```

Comment:
```suggestion
The `max_tombstone_block` is there both to allow that some of the required outputs can be fog outputs (since fog imposes tombstone block limits),
```